### ISSUE TITLE
3.0.103: No space left on device

### ISSUE THREAD

**Issue by onlyjob** [moosefs/moosefs#201]
For no apparent reason, sometimes I'm getting "No space left on device" errors when writing data to MooseFS, despite plenty of space on all HDDs...

**Comment by xandrus** [comment]
Would you be so kind and tell us something more about this issue?
Do you use storage class with strict chunk creation mode?
How many hard disk and chunkservers you have?

By the way what kind of filesystem you use for chunkserver disks?

**Comment by onlyjob** [comment]
File system is _ext4_ (although it is irrelevant). Storage class is 

    -m S -C C,C -K P,P two

There are two `C` chunkservers, each with 4 SSDs.
There are three `P` chunkservers with rotational HDDs.

There are plenty of free space everywhere and I observed the problem with about 20% HDD utilisation in `C`. Chunkservers and HDDs are healthy and I could not see anything of relevance in syslog. The problem is flaky and CGI does not give a clue about what's wrong. Re-trying works for some time until it happens again, once in few hours. There is not much to say about the issue. My _guess_ is that something goes wrong during internal rebalancing and instead of queuing, operation fails on bogus out-of-space error. Several hours ago I've changed _creation mode_ to `LOOSE` and haven't seen the problem since.

Basically, `STRICT` mode shouldn't fail randomly when there are enough healthy chunkservers...

**Comment by zcalusic** [comment]
If it happens only occassionaly, maybe your chunkservers get disconnected occasionaly, due to some network glitches? With strict mode, and rather full disks, at those points in time master would refuse writes, until all chunkservers are available again. Consult master & chunkserver logs, they're useful to detect this behaviour.

Other than that, check that you also have enough available inodes on all underlying file systems. File systems typically return the same ENOSPACE error if you run out of free blocks or free inodes. MooseFS can't really make a distinction between the two.

**Comment by acid-maker** [comment]
If you write a lot of data at the same time to your "C" chunkservers that it is likely that they are overloaded (check 'servers' tab on CGI - 'load' column) and this is why master doesn't want to write data there. Of course in such case you should not get ENOSPC - I'll check that.

**Comment by acid-maker** [comment]
I've just checked the code and it looks ok to me, so we will try to reproduce this on our environement. Maybe we will ask you for more info later.

**Comment by onlyjob** [comment]
Thanks. There were several processes actively writing and I have seen word `OVERLOADED` in "Resources" tab but only momentarily...
Still it is unexpected to get out of space error when chunkserver is merely busy...
Busy chunkserver is a normal operating situation, isn't it? ;)

I've experienced this on new/young cluster with less than 10_000 chunks per HDD on chunkservers 'C' (and around 100_000 chunks per HDD on chunkservers `P`). This is definitely not a lack of inodes.
Network is very stable, all nodes are on two stacked 10 GbE switches within 10 meters from each other.

**Comment by acid-maker** [comment]
Yes. According to the source code (and logic of course) even when chunkservers are overloaded you shouldn't get ENOSPC. In such case mfsmount should get something like EAGAIN and should repeat such write after a short delay. This looks like a very rare bug because you are probably the first person that is experiencing that. Still we need to investigate this.

**Comment by acid-maker** [comment]
I have a question. Do you see anything unusual in your syslog on mfsmount or mfsmaster side?

**Comment by onlyjob** [comment]
From the first look, nothing whatsoever. I'll check the logs again while trying to reproduce the problem.
As for rare bug, I've triggered it so easily, multiple times... :(

**Comment by onlyjob** [comment]
Also, file system is mounted with `mfsmemlock,mfsioretries=444` and memory locking is enabled on all chunkservers.

**Comment by onlyjob** [comment]
Reproduced: it looks like all it takes to trigger out-of-space condition is network glitch. On a hunch, while copying data to _mfs_, I've rebooted a switch between `C/P` pairs of chunkservers and error followed almost immediately:

~~~~
mfschunkserver[117973]: connect to {CHUNKSERVER} failed, error: ETIMEDOUT (Operation timed out)

mfsmount[89510]: file: 442696, index: 200 - fs_writechunk returned status: No space left
mfsmount[89510]: error writing file number 442696: ENOSPC (No space left on device)
~~~~

Ideally `mfsmount` should retry up to number of `mfsioretries` so (for example) restart of chunkserver or temporary connectivity loss would not abort an I/O operation.

**Comment by acid-maker** [comment]
I've just commited small fix. I hope that it helps.

**Comment by oxide94** [comment]
Sorry, I accidentally referenced the wrong commit. Please omit the last reference.
